Enable Microsoft.CodeAnalysis.NetAnalyzers code analyzers

xtqqczze opened the pull request:
# PR Summary

Minimal first step to enable code analyzers in build and live analysis.

* Add `Analyzers.props`
* Add `CodeAnalysis.ruleset` with rules copied from `microsoft.codeanalysis.netanalyzers\5.0.0-rtm.20502.2\rulesets\AllRulesDefault.ruleset`

## PR Context

see also: #11916

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [NA] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [NA] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [NA] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [NA] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

iSazonov commented:
In #11916 my intention is to force Roslyn analyzers at build time to exclude regressions in code formatting and code style.

Here you add analyzers but not enable/activate them for build or design time (otherwise the build would fail). Also we already have editorconfig file - why do we need the ruleset. I am not sure that we need this in the time.

It would great if you grabbed  #11916 and improved it.

xtqqczze commented:
@iSazonov  yes this PR has a similar intention as #11916, but review should be easier 😛

Please see 6acc529 for proof analyzers are activated.

iSazonov commented:
@xtqqczze Thanks for clarify! I see the behavior of VS was changed since 16.5. But It is not clear has EditorConfig a priority on CodeAnalysisRuleSet?

iSazonov commented on Analyzers.props:
```diff
@@ -0,0 +1,9 @@
+<Project>
```
Maybe rename to CodeAnalysis.props?

xtqqczze replied:
the name is from <https://github.com/dotnet/runtime/blob/master/eng/Analyzers.props> but I can change if you would like?

iSazonov replied:
My thought is that the name like `CodeAnalysis.ruleset`.

xtqqczze replied:
My preference is `Analyzers.props` for consistency with both dotnet/runtime repo and the suffix from analyzer package names, i.e. `microsoft.codeanalysis.analyzers`, `stylecop.analyzers`, `xunit.analyzers`

adityapatwardhan commented:
@xtqqczze  Please have a look at the failing builds.

iSazonov commented:
@xtqqczze Could you please check with CA1001 that EditorConfig has a priority over CodeAnalysisRuleSet?

xtqqczze commented:
@iSazonov  2c736cb shows EditorConfig has a priority over CodeAnalysisRuleSet

iSazonov commented:
In the case I suggest to use default ruleset and do all customizations in EditorConfig.

adityapatwardhan commented:
@xtqqczze Can respond to @iSazonov comment 
> In the case I suggest to use default ruleset and do all customizations in EditorConfig.

xtqqczze commented:
> 
> 
> In the case I suggest to use default ruleset and do all customizations in EditorConfig.

I think using Ruleset only makes more sense than specifiying default rules in Ruleset and overriding some rules in EditorConfig.

xtqqczze commented:
@iSazonov I think we should use a Ruleset until EditorConfig support for diagnostics is mature.

iSazonov commented:
TreatWarningsAsErrors  ike a very rare case. We don't use this.

iSazonov commented:
@xtqqczze It seems the https://github.com/dotnet/roslyn/issues/43051 issue consider a specific scenario to have an exclusion list of warnings. It is not our case. We use TreatWarningsAsErrors to force all warning as errors.

iSazonov commented:
/apz restart

iSazonov commented:
@xtqqczze Please rebase. I hope this will fix CIs.

iSazonov commented:
> Ruleset based on microsoft.codeanalysis.fxcopanalyzers\3.0.0\rulesets\AllRulesDefault.ruleset with actions changed from Warning to Info to fix the build.

Please enumerate these changes. Maybe in new issue. This helps us to fix its in follow PRs.

iSazonov commented:
@xtqqczze Please add comments in rules so that we fix temporary changes in follow PRs.

iSazonov commented:
See https://github.com/dotnet/runtime/pull/42272#pullrequestreview-488980943 about RunAnalyzers

iSazonov commented:
@xtqqczze I see .Net team updated their code (rule list and csproj) - could you please update the PR too?

xtqqczze commented:
rebased

xtqqczze commented:
@iSazonov I think it would make more sense to abandon this PR, and start from scratch with new `Microsoft.CodeAnalysis.NetAnalyzers` using EditorConfig rules.

iSazonov commented:
> @iSazonov I think it would make more sense to abandon this PR, and start from scratch with new `Microsoft.CodeAnalysis.NetAnalyzers` using EditorConfig rules.

Ok.

xtqqczze commented:
> > @iSazonov I think it would make more sense to abandon this PR, and start from scratch with new `Microsoft.CodeAnalysis.NetAnalyzers` using EditorConfig rules.
> 
> Ok.

> > Previously, you had wanted to use EditorConfig instead of ruleset configuration.
> 
> My thought was and is that we shouldn't have two configs - in EditorConfig and in ruleset. If VS Code supports(?) ruleset it is better to use ruleset and follow .Net team's pattern.
> 
> _Originally posted by @iSazonov in https://github.com/PowerShell/PowerShell/pull/13835#issuecomment-714652241_

Let's stick with ruleset configuration then. Honestly, I don't have a preference either way, but I would like to get something merged soon.

xtqqczze committed:
Add Analyzers.props

xtqqczze committed:
Add default ruleset

Based on `\microsoft.codeanalysis.netanalyzers\5.0.0-rc2.20460.4\rulesets\AllRulesDefault.ruleset`

xtqqczze commented:
This PR is ready for review and is the prerequisite enabling further code analyzers so speedy review would be appreciated.

iSazonov commented:
@xtqqczze I see MSFT updated documentation.
https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview?view=vs-2019#build-errors
> Create a .NET 5.0 project which includes analyzers by default in the .NET SDK. Code analysis is enabled, by default, for projects that target .NET 5.0 or later. You can enable code analysis on projects that target earlier .NET versions by setting the EnableNETAnalyzers property to true.

So we already got this.

https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#convert-an-existing-ruleset-file-to-editorconfig-file
> Starting in Visual Studio 2019 version 16.5, ruleset files are deprecated in favor of the EditorConfig file for analyzer configuration for managed code. Most of the Visual Studio tooling for analyzer rule severity configuration has been updated to work on EditorConfig files instead of ruleset files. EditorConfig files allow you to configure both analyzer rule severities and analyzer options, including Visual Studio IDE code style options. **It is highly recommended that you convert your existing ruleset file to an EditorConfig file.** It is also recommended that you save the EditorConfig file at the root of your repo or in the solution folder. By using the root of your repo or solution folder, you make sure that the severity settings from this file are automatically applied to the entire repo or solution, respectively.

I think we could start with this recommendation. Sorry that I stopped your new PR - it was right direction.

iSazonov commented:
Please also see https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview#code-style-analysis
about EnforceCodeStyleInBuild. I think we could add this.

xtqqczze committed:
Remove unnecessary PackageReference

xtqqczze committed:
test! should break build

xtqqczze commented:
> So We have no need to reference the package.
I have removed the PackageReference, and pushed commit be5c5c3 which proves that a change in the ruleset will break the build.

xtqqczze commented:
> - In VS2019 16.8, you should be able to workaround this issue by using global analyzer configs as per https://github.com/dotnet/roslyn/issues/43051#issuecomment-692837489.
> - In VS2019 16.9 Preview1 and later, the original reported issue should be resolved and you can use editorconfig file itself and /warnaserror will escalate diagnostics configured as warnings in editorconfig to be errors.
> 
> _Originally posted by @mavasani in https://github.com/dotnet/roslyn/issues/43051#issuecomment-701702375_

I have tested EditorConfig configuration with TreatWarningsAsErrors and it is still not working with VS2019 16.8.0 Preview 5.0. The alternative is to use a global AnalyzerConfig (see #13835).

iSazonov commented:
@xtqqczze TreatWarningsAsErrors issue is about using warning severity only but we should use error severity if we want to force a rule. Yes?

xtqqczze commented:
> @xtqqczze TreatWarningsAsErrors issue is about using warning severity only but we should use error severity if we want to force a rule. Yes?

That is a goal of this PR, to enforce various rules during build.

xtqqczze commented:
> I mean why would we use warning severity if we want to stop build with error?

If `TreatWarningsAsErrors` is enabled then warnings should stop the build.

iSazonov commented:
Since we get a fix for TreatWarningsAsErrors in near future I believe we have no need a workaround today. In EditorConfig we can set severity to error if we want to force a rule.

xtqqczze commented:
> Since we get a fix for TreatWarningsAsErrors in near future I believe we have no need a workaround today. In EditorConfig we can set severity to error if we want to force a rule.

@iSazonov In that case, I suggest closing this PR in favour of #13835. Can we make a decision please.

xtqqczze commented:
We can use `.globalconfig` file instead, closing.

iSazonov commented:
@xtqqczze Why would we need a global config? We can use our .editorconfig file. If we set a severity of a rule to error we will stop the build if the rule violates. Yes?

xtqqczze commented:
> @xtqqczze Why would we need a global config? We can use our .editorconfig file. If we set a severity of a rule to error we will stop the build if the rule violates. Yes?

Using a global AnalyzerConfig avoid bloating EditorConfig with analyzer configuration. Documentation is at https://docs.microsoft.com/dotnet/fundamentals/code-analysis/configuration-files#naming

iSazonov commented:
> avoids bloating EditorConfig with analyzer configuration

Bloating? Our file has 194 lines and 142 lines of them with analyzer configuration. I don't expect we will add a lot of rules in near time (maybe ~10 to force new syntax). I believe can live with the EditorConfig while.

xtqqczze commented:
> > avoids bloating EditorConfig with analyzer configuration
> 
> Bloating? Our file has 194 lines and 142 lines of them with analyzer configuration. I don't expect we will add a lot of rules in near time (maybe ~10 to force new syntax). I believe can live with the EditorConfig while. Also global config does not still supported in VS GUI that is bad UX.

I think we would want to specify the severity for each rule explicitly as in #13835. There are 246 rules in `Microsoft.CodeAnalysis.NetAnalyzers` alone, ~738 additional lines.

My main concern  analyzer configuration options

iSazonov commented:
> I think we would want to specify the severity for each rule explicitly (like #13835). There are 246 rules in Microsoft.CodeAnalysis.NetAnalyzers alone, ~738 additional lines.

Currently we use defaults and it is well. Do you expect that we will change defaults and set severity to error for 246 rules?

> AnalyzerConfig is a simple text file, I'm not sure what IDE support is needed.

It is mandatory requirement that IDE support Roslyn analyzer configuration. I know VS Code supports EditorConfig but I am not sure it supports GlobalConfig. Can you confirm?

iSazonov commented:
I tried VS Code with GlobalConfig and it does not work at all :-( OmniSharp uses defaults and ignore the global config. I will open the feature request in OmniSharp repo.

@xtqqczze Now I suggest to put all global config content to a end of out EditorConfig file. After we get the feature in OmniSharp we will be able to move the rules back to a global config file.

xtqqczze commented:
> OmniSharp

Works for me. Do you have `omnisharp.enableRoslynAnalyzers` set to `true` in `.vscode\settings.json`

iSazonov commented:
@xtqqczze EditorConfig works but GlobalConfig doesn't work.

xtqqczze commented:
@iSazonov I tested in VS Code 1.50.1 (Omnisharp 1.37.3) with `src\Microsoft.PowerShell.Commands.Management\cimSupport\cmdletization\SessionBasedWrapper.cs` open in the editor.

As expected, a diagnostic with code CA1507 is listed in the problems tab.

Then I set `dotnet_diagnostic.CA1507.severity = silent` in `.globalconfig` and restarted VS Code and **the diagnostic is no longer listed**.

In my environment, OmniSharp is using the MSBuild instance provided by `Visual Studio Enterprise 2019 16.8.30615.102`.

Perhaps you could you try setting `omnisharp.loggingLevel` to `debug` and examine the log output?

iSazonov commented:
> Visual Studio Enterprise 2019 16.8.30615.102

I updated to latest preview and now it works. Thanks!

xtqqczze commented:
> It is mandatory requirement that IDE support Roslyn analyzer configuration. I know VS Code supports EditorConfig but I am not sure it supports GlobalConfig. Can you confirm?

Support for `.globalconfig` is built into roslyn, so it should work in any IDE (including Visual Studio Enterprise 2019 16.8)
